cmd/end2endtest/ballot.go

```diff
@@ -7,104 +7,248 @@ import (
+func (t *E2EBallotRange) Setup(api *apiclient.HTTPclient, c *config) error {
 	t.api = api
 	t.config = c
 
-	p := &models.Process{
-		StartBlock: 0,
-		BlockCount: 100,
-		Status:     models.ProcessStatus_READY,
-		EnvelopeType: &models.EnvelopeType{
-			EncryptedVotes: false,
-			UniqueValues:   true},
-		CensusOrigin: models.CensusOrigin_OFF_CHAIN_TREE_WEIGHTED,
-		VoteOptions: &models.ProcessVoteOptions{
-			MaxCount:     2,
-			MaxValue:     6,
-			MaxTotalCost: 10,
-			CostExponent: 1,
-		},
-		Mode: &models.ProcessMode{
-			AutoStart:     true,
-			Interruptible: true,
-		},
-		MaxCensusSize: uint64(t.config.nvotes),
+	// setup for range voting
+	p := newTestProcess()
+	p.VoteOptions = &models.ProcessVoteOptions{
+		MaxCount:     4,
+		MaxValue:     3,
+		MaxTotalCost: 10,
+		CostExponent: 1,
 	}
 
 	if err := t.setupElectionRaw(p); err != nil {
+		return fmt.Errorf("error in setupElectionRaw for range voting: %s", err)
+	}
+
+	log.Debugf("election details: %+v", *t.election)
+	return nil
+}
+
+func (*E2EBallotRange) Teardown() error {
+	// nothing to do here
+	return nil
+}
+
+func (t *E2EBallotRange) Run() error {
+	nvotes := t.config.nvotes
+
+	choices, expectedResults := ballotVotes(
+		t.election.TallyMode, nvotes,
+		rangeVote, t.election.VoteMode.UniqueValues)
+
+	if err := sendAndValidateVotes(t.e2eElection, choices, expectedResults); err != nil {
 		return err
 	}
 
+	return nil
+}
+
+func (t *E2EBallotApproval) Setup(api *apiclient.HTTPclient, c *config) error {
+	t.api = api
+	t.config = c
+
+	//setup for approval voting
+	p := newTestProcess()
+	// update uniqueValues to false
+	p.EnvelopeType.UniqueValues = false
+	p.VoteOptions = &models.ProcessVoteOptions{
+		MaxCount:     4,
+		MaxValue:     1,
+		MaxTotalCost: 4,
+		CostExponent: 1,
+	}
+
+	if err := t.setupElectionRaw(p); err != nil {
+		return fmt.Errorf("error in setupElectionRaw for approval voting: %s", err)
+	}
+
 	log.Debugf("election details: %+v", *t.election)
 	return nil
 }
 
-func (*E2EBallotElection) Teardown() error {
+func (*E2EBallotApproval) Teardown() error {
 	// nothing to do here
 	return nil
 }
 
-func (t *E2EBallotElection) Run() error {
-	bdata := ballotData{
-		maxValue:     t.election.TallyMode.MaxValue,
-		maxCount:     t.election.TallyMode.MaxCount,
-		maxTotalCost: t.election.TallyMode.MaxTotalCost,
-		costExponent: t.election.TallyMode.CostExponent,
+func (t *E2EBallotApproval) Run() error {
+	nvotes := t.config.nvotes
+
+	choices, expectedResults := ballotVotes(
+		t.election.TallyMode, nvotes,
+		approvalVote, t.election.VoteMode.UniqueValues)
+
+	if err := sendAndValidateVotes(t.e2eElection, choices, expectedResults); err != nil {
+		return err
 	}
 
-	choices, expectedResults := ballotVotes(bdata, len(t.voterAccounts))
+	return nil
+}
 
-	// Send the votes (parallelized)
+func sendAndValidateVotes(e e2eElection, choices [][]int, expectedResults [][]*types.BigInt) error {
```

Comment:
would this func be useful to other tests? i.e. is it generic enough to be reused elsewhere?
if so, please move it to helpers.go (even if not used right now by any other tests than the ones in ballot.go)